spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/ConfigUtils.java

```diff
@@ -251,12 +269,156 @@ public static boolean noSources(List<?> sources, String namespace) {
 		return false;
 	}
 
+	public static LinkedHashSet<StrictProfile> profiles(boolean includeProfileSpecificSources,
+			LinkedHashSet<String> strictForProfiles, Environment environment) {
+
+		String[] activeProfiles = environment.getActiveProfiles();
+		Set<String> activeProfilesAsSet = Arrays.stream(activeProfiles).collect(Collectors.toSet());
+
+		if (!includeProfileSpecificSources) {
+			LOG.debug("include-profile-specific-sources is false, thus ignoring strict-for-profiles");
+			return new LinkedHashSet<>();
+		}
+
+		if (strictForProfiles.isEmpty()) {
+			LOG.debug("include-profile-specific-sources = true, strict-for-profiles = empty");
+			return StrictProfile.allNonStrict(activeProfiles);
+		}
+
+		// include-profile-specific-sources = true, strict-for-profiles != empty
+		// 1. make sure that strict-for-profiles are all present
+		for (String profile : strictForProfiles) {
+			boolean isPresent = activeProfilesAsSet.contains(profile);
+			if (!isPresent) {
+				throw new StrictSourceNotFoundException("profile : " + profile + " is not an active profile, but there "
+						+ "is source definition for it under 'strict-for-profiles'");
+			}
+		}
+
+		// 2. non-strict profiles are supposed to come first, since we treat them as
+		// "optional"
+		LinkedHashSet<StrictProfile> result = new LinkedHashSet<>();
+		for (String profile : activeProfiles) {
+			if (!strictForProfiles.contains(profile)) {
+				result.add(new StrictProfile(profile, false));
+			}
+		}
+
+		// 3. strict-for-profiles will come in strict order
+		result.addAll(StrictProfile.allStrict(strictForProfiles));
+		return result;
+
+	}
+
+	/**
+	 * <pre>
+	 * 	 root sources are the ones that do _not_ end in profile.name from profiles.
+	 * 	 for example, we have StrictProfiles : {[dev, false], [prod, true]} and byLabels:
+	 * 	 - a-dev
+	 * 	 - b-prod
+	 * 	 - c
+	 *
+	 * 	 in such a case, rootProfiles are considered only : "c".
+	 *
+	 * </pre>
+	 */
+	static List<StrippedSourceContainer> rootSources(List<StrippedSourceContainer> byLabels,
```

Comment:
- I find everything in the namespace by labels
- `profiles` tracks all profiles - both strict and non-strict.
- I iterate all sources and check if there is subset of them, that do not have "-<PROFILE>" appended to them => these are root sources. For example:

```
byLabels = {a, a-dev, b, b-prod}
profiles = {[dev, false], [prod, true]}
```

from such a set-up, `a` and `b` are going to be selected as root sources.